File: lab_2/sort.py
import random

fin = open('sort.in', 'r')
fout = open('sort.out', 'w')

data = fin.read().strip().split('\n')
numbers_raw = data[1].split()

numbers = [int(i) for i in numbers_raw if i.strip('-').isnumeric()]

def quicksort(array: list) -> list:
    if len(array) < 2:
        return array
    else:
        index = random.randint(0, len(array) - 1)
        pivot = array[index]
        array.pop(index)
        less = [i for i in array if i <= pivot]
        greater = [i for i in array if i > pivot]

        result = quicksort(less) + [pivot] + quicksort(greater)
        return result

r = quicksort(numbers)
print(' '.join([str(i) for i in r]), file = fout)

File: lab_2/sort.go
package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

func quickSort(array []int) []int {
	n := len(array)
	if n < 2 {
		return array
	} else {
		index := rand.Intn(n)
		pivot := array[index]
		var less []int
		var greater []int

		for i := 0; i < n; i++ {
			if i != index {
				if array[i] <= pivot {
					less = append(less, array[i])
				} else {
					greater = append(greater, array[i])
				}
			}
		}

		result := append(quickSort(less), pivot)
		result = append(result, quickSort(greater)...)
		return result
	}
}

func main() {
	var n int
	data_raw, _ := ioutil.ReadFile("sort.in")
	data := strings.Split(string(data_raw), "\n")
	fmt.Sscanf(data[0], "%d", &n)
	array := make([]int, n)
	numbersRaw := strings.Split(data[1], " ")

	for i := 0; i < n; i++ {
		array[i], _ = strconv.Atoi(numbersRaw[i])
	}

	sorted := quickSort(array)
	resultRaw := make([]string, n)

	for i := 0; i < n; i++ {
		resultRaw[i] = fmt.Sprint(sorted[i])
	}

	result := strings.Join(resultRaw, " ")
	fout, _ := os.Create("sort.out")
	fout.WriteString(result)
	fout.Close()
}
